src/box/gc.c

```diff
@@ -733,6 +806,280 @@ gc_consumer_name(struct gc_consumer *consumer)
 			  tt_uuid_str(&consumer->uuid));
 }
 
+/**
+ * Write the consumer to persistent WAL GC state. The function must be called
+ * only when schema supports persistent consumers. If consumer wasn't in space
+ * _gc_consumers, the new row is created.
+ */
+static int
+gc_consumer_persist_impl(struct gc_consumer *consumer)
+{
+	assert(gc_schema_supports_persistent_consumers());
+
+	/* Mark consumer as synchronized with persistent state. */
+	consumer->is_synced = true;
+	char tuple_buf[UUID_STR_LEN + VCLOCK_STR_LEN_MAX + GC_NAME_MAX + 30];
+	char *tuple_end = tuple_buf;
+	tuple_end = mp_encode_array(tuple_end, 3);
+	tuple_end = mp_encode_str0(tuple_end, tt_uuid_str(&consumer->uuid));
+	tuple_end = mp_encode_vclock_ignore0(tuple_end, &consumer->vclock);
+	tuple_end = mp_encode_map(tuple_end, 1);
+	tuple_end = mp_encode_str0(tuple_end, "type");
+	tuple_end = mp_encode_str0(tuple_end,
+				   gc_consumer_type_strs[consumer->type]);
+	assert((unsigned long)(tuple_end - tuple_buf) < sizeof(tuple_buf));
+
+	return box_replace(BOX_GC_CONSUMERS_ID, tuple_buf, tuple_end, NULL);
+}
+
+int
+gc_consumer_persist(struct gc_consumer *consumer)
+{
+	if (gc_schema_supports_persistent_consumers()) {
+		struct txn *txn = txn_begin();
+		if (txn == NULL)
+			return -1;
+		/*
+		 * No need to wait limbo, so let's set TXN_FORCE_ASYNC.
+		 * Moreover, the function can be called when quorum cannot be
+		 * collected, (for example, on IPROTO_SUBSCRIBE), so we must set
+		 * this flag here.
+		 */
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (gc_consumer_persist_impl(consumer) != 0) {
+			txn_abort(txn);
+			return -1;
+		}
+		if (txn_commit(txn) != 0)
+			return -1;
+	}
+	/* Unref WAL GC delay if being persisted for the first time. */
+	if (!consumer->is_persistent)
+		gc_delay_unref();
+	consumer->is_persistent = true;
```

Comment:
Shouldn't we set this flag only if the schema supports persistent consumers?

Reply:
Wrote an explanatory comment here.